File: Triangle.java
public class Solution {
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int n = triangle.size();
        int []dp = new int[n+1];   
        for(int i = n-1; i > -1; i--)
        {      
            for(int j=0; j < triangle.get(i).size(); j++)
            {
                int val = (int)triangle.get(i).get(j);
                dp[j] = val + Math.min(dp[j], dp[j+1]);
            }
        }
        return dp[0];
    }
}

File: Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int> > &triangle) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<int> row(triangle.size()+1, 0);
        
        const int R = triangle.size();
        for(int i = R-1; i>-1; i--)
        {
            const int C = triangle[i].size();
            for(int j = 0; j < C; j++)
            {
                row[j] = row[j]<row[j+1]? row[j]:row[j+1];
                row[j] += triangle[i][j];
            }
        }
        
        
        return row[0];
    }
};
